presto-main/src/main/java/io/prestosql/server/security/oauth2/OAuth2Service.java

```diff
@@ -0,0 +1,184 @@
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.jsonwebtoken.SigningKeyResolver;
+import io.prestosql.server.security.oauth2.OAuth2Client.AccessToken;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.net.URI;
+import java.security.SecureRandom;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.time.temporal.TemporalAmount;
+import java.util.Date;
+
+import static com.google.common.base.Strings.nullToEmpty;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.hash.Hashing.sha256;
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static java.util.Objects.requireNonNull;
+
+public class OAuth2Service
+{
+    private static final String STATE_AUDIENCE = "presto_oauth";
+    private static final String FAILURE_REPLACEMENT_TEXT = "<!-- ERROR_MESSAGE -->";
+
+    private final OAuth2Client client;
+    private final JwtParser jwtParser;
+
+    private final String failureHtml;
+
+    private final TemporalAmount challengeTimeout;
+    private final byte[] stateHmac;
+
+    @Inject
+    public OAuth2Service(OAuth2Client client, @ForOAuth2 SigningKeyResolver signingKeyResolver, OAuth2Config oauth2Config)
+            throws IOException
+    {
+        this.client = requireNonNull(client, "client is null");
+        this.jwtParser = Jwts.parser().setSigningKeyResolver(signingKeyResolver);
+
+        this.failureHtml = Resources.toString(Resources.getResource(getClass(), "/oauth2/failure.html"), UTF_8);
+        verify(failureHtml.contains(FAILURE_REPLACEMENT_TEXT), "login.html does not contain the replacement text");
+
+        requireNonNull(oauth2Config, "oauth2Config is null");
+        this.challengeTimeout = Duration.ofMillis(oauth2Config.getChallengeTimeout().toMillis());
+        this.stateHmac = oauth2Config.getStateKey()
+                .map(key -> sha256().hashString(key, UTF_8).asBytes())
+                .orElseGet(() -> secureRandomBytes(32));
+    }
+
+    public URI startChallenge(URI callbackUri)
+    {
+        String state = Jwts.builder()
+                .signWith(SignatureAlgorithm.HS256, stateHmac)
+                .setAudience(STATE_AUDIENCE)
+                .setExpiration(Date.from(ZonedDateTime.now().plus(challengeTimeout).toInstant()))
+                .compact();
+
+        return client.getAuthorizationUri(state, callbackUri);
+    }
+
+    public OAuthResult finishChallenge(String state, String code, URI callbackUri)
+            throws ChallengeFailedException
+    {
+        requireNonNull(callbackUri, "callbackUri is null");
+        requireNonNull(state, "state is null");
+        requireNonNull(code, "code is null");
+
+        Claims stateClaims = parseState(state);
+        if (!STATE_AUDIENCE.equals(stateClaims.getAudience())) {
+            // this is very unlikely, but is a good safety check
+            throw new ChallengeFailedException("Unexpected state audience");
```

Comment:
Include the claim audience in the message? Or log here for debugging? If this does occur, I expect the administrator will want to know the value.